File: apps/web/src/lib/storedMoves.ts
import { Chess } from 'chess.js'

const resultTokens = new Set(['1-0', '0-1', '1/2-1/2', '*', 'live'])

export function parseStoredMoves(value?: string) {
  const source = value?.trim()
  if (!source) return []

  try {
    const chess = new Chess()
    chess.loadPgn(source)
    return chess.history()
  } catch {
    return parseLegacyMoveText(source)
  }
}

function parseLegacyMoveText(source: string) {
  const chess = new Chess()
  const tokens = stripVariations(source)
    .replace(/^\s*\[[^\]]*\]\s*$/gm, ' ')
    .replace(/\{[^}]*\}/g, ' ')
    .replace(/;[^\r\n]*/g, ' ')
    .replace(/\$\d+/g, ' ')
    .split(/\s+/)

  for (const rawToken of tokens) {
    const token = rawToken
      .trim()
      .replace(/^\d+\.(?:\.\.)?/, '')
      .replace(/[!?]+$/g, '')
    if (!token || resultTokens.has(token)) continue

    try {
      if (!chess.move(token)) break
    } catch {
      break
    }
  }

  return chess.history()
}

function stripVariations(value: string) {
  let result = value
  let previous = ''
  while (result !== previous) {
    previous = result
    result = result.replace(/\([^()]*\)/g, ' ')
  }
  return result
}

File: apps/mobile/lib/stored_moves.dart
import 'package:chess/chess.dart' as chess;

const _resultTokens = {'1-0', '0-1', '1/2-1/2', '*', 'live'};

List<String> parseStoredMoves(String? value) {
  final source = value?.trim();
  if (source == null || source.isEmpty) return const [];

  try {
    final game = chess.Chess();
    if (game.load_pgn(source)) {
      return game.getHistory().map((move) => move.toString()).toList();
    }
  } catch (_) {
    // Older rows may contain SAN move text rather than a complete PGN.
  }

  return _parseLegacyMoveText(source);
}

List<String> _parseLegacyMoveText(String source) {
  final game = chess.Chess();
  final tokens = _stripVariations(source)
      .replaceAll(RegExp(r'^\s*\[[^\]]*\]\s*$', multiLine: true), ' ')
      .replaceAll(RegExp(r'\{[^}]*\}'), ' ')
      .replaceAll(RegExp(r';[^\r\n]*'), ' ')
      .replaceAll(RegExp(r'\$\d+'), ' ')
      .split(RegExp(r'\s+'));

  for (final rawToken in tokens) {
    final token = rawToken
        .trim()
        .replaceFirst(RegExp(r'^\d+\.(\.\.)?'), '')
        .replaceFirst(RegExp(r'[!?]+$'), '');
    if (token.isEmpty || _resultTokens.contains(token)) continue;

    try {
      if (!game.move(token)) break;
    } catch (_) {
      break;
    }
  }

  return game.getHistory().map((move) => move.toString()).toList();
}

String _stripVariations(String source) {
  var result = source;
  var previous = '';
  while (result != previous) {
    previous = result;
    result = result.replaceAll(RegExp(r'\([^()]*\)'), ' ');
  }
  return result;
}
